src/lsm/compaction.zig

```diff
@@ -281,1843 +302,1648 @@ pub fn CompactionType(
+        // example, only a single value block from each level is needed to start a merge.
         //
-        // At the moment, the forest won't pipeline different compactions from other tree-levels
-        // together. It _can_ do this, but it requires a bit more thought in how memory is managed.
+        // The job of compaction_dispatch is to kick off all the jobs. There are several additional
+        // concerns:
+        // - All jobs use the same common pool of resources (ResourcePool). The jobs are started
+        //   in the order that splits resources fairly (e.g., reads from level a and level b
+        //   alternate). Fairness also ensures that the process does not deadlock.
+        // - Jobs have dependencies --- merging needs value blocks, reading a value block needs the
+        //   corresponding index blocks.
+        // - A single bar of compaction should process only a fraction of the input, so the
+        //   processes can be suspended in the middle.
         //
-        // IO work is always submitted to the kernel _before_ entering blip_merge().
+        // A beat of compaction ends when both:
+        //   - at least quota.bar of input values is consumed,
+        //   - there's no incomplete output value blocks.
         //
-        // TODO: even without a threadpool, we can likely drive better performance by doubling up
-        // the stages. The reason for this is that we expect blip_merge() to be quite a bit quicker
-        // than blip_write().
+        // In other words, the only compaction state that gets carried over to the next beat is a
+        // partially full index block. The current beat must end with writing a data block, and the
+        // next beat must start with re-reading level_a and level_b index and value blocks.
+        fn compaction_dispatch(compaction: *Compaction) void {
+            switch (compaction.stage) {
+                .beat,
+                .beat_quota_done,
+                .beat_table_done,
+                => {},
+                .inactive,
+                .paused,
+                => unreachable,
+            }
 
-        /// Perform read IO to fill our source_index_block_{a,b} and source_value_blocks with as
-        /// many blocks as we can, given their sizes, and where we are in the amount of work we
-        /// need to do this beat.
-        pub fn blip_read(compaction: *Compaction, callback: BlipCallback, ptr: *anyopaque) void {
-            log.debug("blip_read({s}): scheduling read IO", .{compaction.tree_config.name});
+            // The loop below runs while (!progressed) and, every time progressed is set to true,
+            // one of the safety_counter resources is acquired.
+            var progressed = true;
+            const safety_counter =
+                compaction.pool.?.reads.total() +
+                compaction.pool.?.writes.total() +
+                compaction.pool.?.cpus.total() + 1;
```

Comment:
Or maybe even
```suggestion
                compaction.pool.?.reads.available() +
                compaction.pool.?.writes.available() +
                compaction.pool.?.cpus.available() + 1;
```

Reply:
Hm, I was mentally trying to convince you that `total` is better, but ended up convincing myself that `available` is the way to go 🤣 

I can't put my finger on the exact guideline here, it seems the two things are in tension here:

* we want to be our safety counters to be simple, so it is obvious that they are correct.
* but we also want them to be tight, such that, if they are _not_ correct, the VOPR tells us this. 

Here, total is less tight, but simpler, but I think it's not _that_ simpler than `.available`, and that better captures the reason _why_ this is valid safety counter.